File: 204.count-primes.py
#
# @lc app=leetcode id=204 lang=python3
#
# [204] Count Primes
#

# Time:  O(nlog(logn))
# Space: O(n)

# @lc code=start
class Solution:
    def countPrimes(self, n: int) -> int:
        res = 0
        primes = [True] * n
        for i in range(2, n):
            if not primes[i]:
                continue
            res, j = res + 1, 2
            while i * j < n:
                primes[i * j] = False
                j += 1
        return res

        # faster but time complexity doesn't change
        if n < 3:
            return 0

        primes = [False, False] + [True] * (n - 2)
        for i in range(4, n, 2):
            primes[i] = False
        for i in range(3, int(n ** 0.5) + 1, 2):
            if not primes[i]:
                continue
            j = 3
            while i * j < n:
                primes[i * j] = False
                j += 2
        return sum(primes)


# @lc code=end

File: 204.count-primes.go
/*
 * @lc app=leetcode id=204 lang=golang
 *
 * [204] Count Primes
 */

package leetcode

// @lc code=start
func countPrimes(n int) int {
	res, primes := 0, make([]bool, n)
	for i := 0; i < n; i++ {
		primes[i] = true
	}
	for i := 2; i < n; i++ {
		if !primes[i] {
			continue
		}
		res++
		for j := 2; i*j < n; j++ {
			primes[i*j] = false
		}
	}
	return res
}

// @lc code=end
